File: src/game/node-registry.ts
// === Built-in Logic Nodes ===
//
// Every node is a single self-linking unit:
//   condition fulfilled → effect runs → self queued → waits again.
// No gate→show pairs. No special flags.
//
// The clock test demonstrates a pure-predicate condition slot:
//   { check: (bus) => ... } — any boolean formula, not tied to a tag.

import {
	EventTag,
	type ShowDialogEvent,
} from './event-types';
import {type LogicNode, createNode} from './logic-nodes';
import {buildEncounterTable} from './plot/encounters';

// ── Built-in nodes ──

export function createBuiltinNodes(): LogicNode[] {
	const encounters = buildEncounterTable();

	return [
		// Random encounter: condition + effect in one node, self-links
		createNode({
			id: 'enc_random',
			label: 'Random Encounter',
			conditions: [{
				tag: EventTag.PlayerMove,
				predicate(ev) {
					const steps = 'steps' in ev ? (ev as {steps: number}).steps : 0;
					if (steps < 15) {
						return false;
					}

					const chance = 0.01 + Math.min((steps - 15) * 0.001, 0.11);
					return Math.random() < chance;
				},
			}],
			mask: [1],
			next: ['enc_random'],
			tags: ['encounter'],
			effect(ctx) {
				const def = encounters[Math.floor(Math.random() * encounters.length)];
				const dialog: ShowDialogEvent = {
					tag: EventTag.ShowDialog,
					title: def.title,
					description: def.description,
					choices: def.choices,
				};
				ctx.bus.emit(dialog);
			},
		}),

		// Level up: single node, self-links
		createNode({
			id: 'sys_level_up',
			label: 'Level Up',
			conditions: [{tag: EventTag.PlayerLevelUp}],
			mask: [1],
			next: ['sys_level_up'],
			tags: ['system'],
			effect(ctx) {
				const ev = ctx.bus.lastTickEvents.find(event => event.tag === EventTag.PlayerLevelUp);
				const level = ev && 'newLevel' in ev
					? (ev as {newLevel: number}).newLevel
					: 0;
				ctx.bus.emit({
					tag: EventTag.ShowDialog,
					title: 'Level Up!',
					description: `You have reached level ${level}! Your abilities grow stronger.`,
					choices: [{label: 'Continue'}],
				});
			},
		}),

		// Settlement greeting: single node, self-links
		createNode({
			id: 'sys_settlement',
			label: 'Settlement Greeting',
			conditions: [{tag: EventTag.PlayerEnterSettlement}],
			mask: [1],
			next: ['sys_settlement'],
			tags: ['system'],
			effect(ctx) {
				const ev = ctx.bus.lastTickEvents.find(event => event.tag === EventTag.PlayerEnterSettlement);
				const name = ev && 'settlementName' in ev
					? (ev as {settlementName: string}).settlementName
					: 'Unknown';
				ctx.bus.emit({
					tag: EventTag.ShowDialog,
					title: `Welcome to ${name}`,
					description: 'The gates open before you. Merchants hawk their wares and guards patrol the walls.',
					choices: [{label: 'Enter'}],
				});
			},
		}),

	];
}

/** Node IDs that should be activated at game start. */
export const INITIAL_ACTIVE_NODES = [
	'enc_random',
	'sys_level_up',
	'sys_settlement',
];

File: src/events/node_registry.cpp
#include "events/node_registry.h"
#include "events/event_types.h"
#include "events/logic_nodes.h"
#include <memory>
#include <string>
#include <utility>

namespace sm {

namespace {

void add_continue_choice(GameEvent& dialog, const char* label) {
    dialog.dialogChoices = std::make_shared<std::vector<DialogChoicePayload>>();
    dialog.dialogChoices->push_back(DialogChoicePayload{label ? label : "Continue", {}, {}});
    dialog.ix = static_cast<int>(dialog.dialogChoices->size());
}

const GameEvent* find_settlement_enter_event(const EventBus& bus) {
    for (const auto& ev : bus.last_tick_events()) {
        if (ev.tag == EventTag::SettlementVisit
            || ev.tag == EventTag::PlayerEnterSettlement) {
            return &ev;
        }
    }
    return nullptr;
}

// An "enc_random" node used to sit here: every 15 walked cells it rolled a
// 1-12% die (from its OWN hardcoded-seed RNG, world-independent) and, on a
// hit, popped a UNIFORMLY RANDOM row of content::encounters() — no biome, no
// danger zone, no faction territory, no time of day, no player state. Removed
// by owner ruling (2026-08-05): events must arise from GAME CONTEXT and
// STATE, never from an unconditional random roll over a list — that trigger
// was the old disease, not a system. The encounter TABLE (content/plot/
// encounters.cpp) and the whole ShowDialog-with-choices → effect_applicator
// path stay: they are the presentation half a future context-driven trigger
// (work_vector №3) will drive.

// A "sys_level_up" node used to sit here, waiting on EventTag::PlayerLevelUp to
// pop a "Level Up!" dialog. Nothing in the project ever emitted that event —
// all four award_exp call sites (effect_applicator, quest_engine, the subworld
// kill path, the addexp console command) take a LevelData and no bus — so the
// node could not fire, and the player has never seen the dialog. Removed rather
// than left as furniture (owner's call, 2026-08-05); the event tag stays in
// event_types.h for whoever wires levelling feedback up for real.

LogicNode settlement_dialog_node() {
    LogicNode n;
    n.id = "sys_settlement";
    n.label = "Settlement Greeting";
    ConditionSlot c;
    c.isEvent = false;
    c.check = [](const EventBus& bus, const PlayerState&) {
        return find_settlement_enter_event(bus) != nullptr;
    };
    n.conditions.push_back(std::move(c));
    n.mask.push_back(1);
    n.next.push_back(n.id);
    n.tags.push_back("system");
    n.effect = [](NodeContext& ctx) {
        std::string name = "Unknown";
        if (const GameEvent* ev = find_settlement_enter_event(*ctx.bus)) {
            if (!ev->s1.empty()) name = ev->s1;
        }

        GameEvent dialog{EventTag::ShowDialog};
        dialog.s1 = "Welcome to " + name;
        dialog.s2 = "The gates open before you. Merchants hawk their wares and guards patrol the walls.";
        add_continue_choice(dialog, "Enter");
        ctx.bus->emit(dialog);
    };
    return n;
}

} // namespace

void register_builtin_nodes(LogicNodeEngine& logic) {
    logic.add(settlement_dialog_node());
    logic.activate("sys_settlement");
}

} // namespace sm
